File: src/pactester.c
// Copyright (C) 2008-2023 Manu Garg.
// Author: Manu Garg <manugarg@gmail.com>
//
// This file implements pactester using pacparser.
//
// pacparser is a library that provides methods to parse proxy auto-config
// (PAC) files. Please read README file included with this package for more
// information about this library.
//
// pacparser is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.

// pacparser is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.

// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA

#include "pacparser.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define STREQ(s1, s2) (strcmp((s1), (s2)) == 0)

#define LINEMAX 4096  // Max length of any line read from text files (4 KiB)
#define PACMAX (1024 * 1024)  // Max size of the PAC script (1 MiB)

__attribute__((noreturn)) void usage(const char *progname)
{
  fprintf(stderr, "\nUsage:  %s <-p pacfile> <-u url> [-h host] "
          "[-c client_ip] [-e]", progname);
  fprintf(stderr, "\n        %s <-p pacfile> <-f urlslist> "
          "[-c client_ip] [-e]\n", progname);
  fprintf(stderr, "\nOptions:\n");
  fprintf(stderr, "  -p pacfile   : PAC file to test (specify '-' to read "
                  "from standard input)\n");
  fprintf(stderr, "  -u url       : URL to test for\n");
  fprintf(stderr, "  -h host      : Host part of the URL\n");
  fprintf(stderr, "  -c client_ip : client IP address (as returned by "
                  "myIpAddres() function\n");
  fprintf(stderr, "                 in PAC files), defaults to IP address "
                  "on which it is running.\n");
  fprintf(stderr, "  -e           : Deprecated: IPv6 extensions are enabled"
                  "by default now.\n");
  fprintf(stderr, "  -f urlslist  : a file containing list of URLs to be "
          "tested.\n");
  fprintf(stderr, "  -v           : print version and exit\n");
  exit(1);
}

char *get_host_from_url(const char *url)
{
  // copy  url to a  pointer that we'll use to seek through the string.
  char *p = strdup(url);
  char *q = p;
  // Move to :
  while (*p != ':' && *p != '\0')
    p++;
  if (p[0] == '\0'||                    // We reached end without hitting :
      p[1] != '/' || p[2] != '/'        // Next two characters are not //
      ) {
    free(q);
    fprintf(stderr, "pactester.c: Not a proper URL\n");
    return NULL;
  }
  p = p + 3;                            // Get past '://'
  // Host part starts from here.
  char *host = p;
  if (*p == '\0' || *p == '/' || *p == ':') {   // If host part is null.
    free(q);
    fprintf(stderr, "pactester.c: Not a proper URL\n");
    return NULL;
  }
  // Seek until next /, : or end of string.
  while (*p != '/' && *p != ':' && *p != '\0')
    p++;
  *p = '\0';
  return host;
}

int main(int argc, char* argv[])
{
  char *pacfile = NULL, *url = NULL, *host = NULL, *urlslist = NULL,
       *client_ip = NULL;

  if (argv[1] && (STREQ(argv[1], "--help") || STREQ(argv[1], "--helpshort"))) {
    usage(argv[0]);
  }

  signed char c;
  while ((c = getopt(argc, argv, "evp:u:h:f:c:")) != -1)
    switch (c)
    {
      case 'v':
        printf("%s\n", pacparser_version());
        return 0;
      case 'p':
        pacfile = optarg;
        break;
      case 'u':
        url = optarg;
        break;
      case 'h':
        host = optarg;
        break;
      case 'f':
        urlslist = optarg;
        break;
      case 'c':
        client_ip = optarg;
        break;
      case 'e':
        break;
      case '?':
        usage(argv[0]);
        /* fallthrough */
      default:
        abort ();
    }

  if (!pacfile) {
    fprintf(stderr, "pactester.c: You didn't specify the PAC file\n");
    usage(argv[0]);
  }
  if (!url && !urlslist) {
    fprintf(stderr, "pactester.c: You didn't specify the URL\n");
    usage(argv[0]);
  }

  // Initialize pacparser.
  if (!pacparser_init()) {
      fprintf(stderr, "pactester.c: Could not initialize pacparser\n");
      return 1;
  }

  // Read pacfile from stdin.
  if (STREQ("-", pacfile)) {
    char *script;
    size_t script_size = 1;  // for the null terminator
    char buffer[LINEMAX];

    script = (char *) malloc(sizeof(char) * LINEMAX);
    if (script == NULL) {
      perror("pactetser.c: Failed to allocate the memory for the script");
      return 1;
    }
    script[0] = '\0';                   // Null terminate to prepare for strcat

    while (fgets(buffer, LINEMAX, stdin)) {
      if (strlen(buffer) == 0)
        break;
      script_size += strlen(buffer);
      if (script_size > PACMAX) {
        fprintf(stderr, "Input file is too big. Maximum allowed size is: %d",
                PACMAX);
        exit(1);
      }
      script = realloc(script, script_size);
      if (script == NULL) {
        perror("pactester.c: Failed to allocate the memory for the script");
        exit(1);
      }
      strcat(script, buffer);
    }

    if (ferror(stdin)) {
      free(script);
      perror("pactester.c: Error reading from stdin");
      return 1;
    }

    if (!pacparser_parse_pac_string(script)) {
      fprintf(stderr, "pactester.c: Could not parse the pac script: %s\n",
              script);
      pacparser_cleanup();
      exit(1);
    }
    free(script);
  }
  else {
    if (!pacparser_parse_pac_file(pacfile)) {
      fprintf(stderr, "pactester.c: Could not parse the pac file: %s\n",
              pacfile);
      pacparser_cleanup();
      exit(1);
    }
  }

  if (client_ip && !pacparser_setmyip(client_ip)) {
    fprintf(stderr, "pactester.c: Error setting client IP\n");
    pacparser_cleanup();
    exit(1);
  }

  char *proxy;

  if (url) {
    // If the host was not explicitly given, get it from the URL.
    // If that fails, return with error (the get_host_from_url()
    // function will print a proper error message in that case).
    host = host ? host: get_host_from_url(url);
    if (!host) {
      exit(1);
    }
    proxy = pacparser_find_proxy(url, host);
    if (proxy == NULL) {
      fprintf(stderr, "pactester.c: %s %s.\n",
              "Problem in finding proxy for", url);
      pacparser_cleanup();
      exit(1);
    }
    printf("%s\n", proxy);
    exit(0);
  }

  if (urlslist) {
    char line[LINEMAX];
    FILE *fp;
    if (!(fp = fopen(urlslist, "r"))) {
      fprintf(stderr, "pactester.c: Could not open urlslist: %s", urlslist);
      pacparser_cleanup();
      exit(1);
    }
    while (fgets(line, sizeof(line), fp)) {
      char *url = line;
      // Remove spaces from the beginning.
      while (*url == ' ' || *url == '\t')
        url++;
      // Skip comment lines.
      if (*url == '#') {
        printf("%s", url);
        continue;
      }
      char *urlend = url;
      while (*urlend != '\r' && *urlend != '\n' && *urlend != '\0' &&
             *urlend != ' ' && *urlend != '\t')
        urlend++;  // keep moving till you hit space or end of string
      *urlend = '\0';
      if (!(host = get_host_from_url(url)) )
        continue;
      proxy = NULL;
      proxy = pacparser_find_proxy(url, host);
      if (proxy == NULL) {
        fprintf(stderr, "pactester.c: %s %s.\n",
                "Problem in finding proxy for", url);
        pacparser_cleanup();
        exit(1);
      }
      if (proxy)
        printf("%s : %s\n", url, proxy);
    }
    fclose(fp);
    exit(0);
  }

  pacparser_cleanup();
  return 0;
}

File: src/pymod/pacparser/pactester.py
# Copyright (C) 2007-2026 Manu Garg.
# Author: Manu Garg <manugarg@gmail.com>
#
# pacparser is a library that provides methods to parse proxy auto-config
# (PAC) files. Please read README file included with this package for more
# information about this library.
#
# pacparser is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.

# pacparser is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.

# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
# USA.

"""
Command line tool to test PAC files. It provides the same interface as the
C pactester tool:

  pactester <-p pacfile> <-u url> [-h host] [-c client_ip] [-e]
  pactester <-p pacfile> <-f urlslist> [-c client_ip] [-e]
"""

import getopt
import sys

import pacparser

PACMAX = 1024 * 1024  # Max size of the PAC script (1 MiB)

PROGNAME = "pactester"

def usage():
  sys.stderr.write("\nUsage:  %s <-p pacfile> <-u url> [-h host] "
                   "[-c client_ip] [-e]\n" % PROGNAME)
  sys.stderr.write("        %s <-p pacfile> <-f urlslist> "
                   "[-c client_ip] [-e]\n\n" % PROGNAME)
  sys.stderr.write("Options:\n")
  sys.stderr.write("  -p pacfile   : PAC file to test (specify '-' to read "
                   "from standard input)\n")
  sys.stderr.write("  -u url       : URL to test for\n")
  sys.stderr.write("  -h host      : Host part of the URL\n")
  sys.stderr.write("  -c client_ip : client IP address (as returned by "
                   "myIpAddres() function\n")
  sys.stderr.write("                 in PAC files), defaults to IP address "
                   "on which it is running.\n")
  sys.stderr.write("  -e           : Deprecated: IPv6 extensions are enabled"
                   "by default now.\n")
  sys.stderr.write("  -f urlslist  : a file containing list of URLs to be "
                   "tested.\n")
  sys.stderr.write("  -v           : print version and exit\n")
  sys.exit(1)

def get_host_from_url(url):
  # Extract the host part of the URL, like the C pactester does. Prints an
  # error and returns None if the URL is not in a recognized form.
  idx = url.find(':')
  if idx < 0 or url[idx + 1:idx + 3] != '//':
    sys.stderr.write("pactester: Not a proper URL\n")
    return None
  start = idx + 3
  if start >= len(url) or url[start] in ('/', ':'):
    sys.stderr.write("pactester: Not a proper URL\n")
    return None
  for end in range(start, len(url)):
    if url[end] in ('/', ':'):
      break
  else:
    end = len(url)
  host = url[start:end]
  if not host:
    sys.stderr.write("pactester: Not a proper URL\n")
    return None
  return host

def main():
  argv = sys.argv[1:]
  if argv and argv[0] in ("--help", "--helpshort"):
    usage()

  pacfile = url = host = urlslist = client_ip = None

  try:
    opts, _ = getopt.getopt(argv, "evp:u:h:f:c:")
  except getopt.GetoptError:
    usage()

  for opt, arg in opts:
    if opt == "-v":
      print(pacparser.version())
      return 0
    elif opt == "-p":
      pacfile = arg
    elif opt == "-u":
      url = arg
    elif opt == "-h":
      host = arg
    elif opt == "-f":
      urlslist = arg
    elif opt == "-c":
      client_ip = arg
    elif opt == "-e":
      pass

  if not pacfile:
    sys.stderr.write("pactester: You didn't specify the PAC file\n")
    usage()
  if not url and not urlslist:
    sys.stderr.write("pactester: You didn't specify the URL\n")
    usage()

  try:
    pacparser.init()
  except Exception:
    sys.stderr.write("pactester: Could not initialize pacparser\n")
    return 1

  if pacfile == "-":
    script = sys.stdin.read()
    if len(script) > PACMAX:
      sys.stderr.write("Input file is too big. Maximum allowed size is: %d"
                       % PACMAX)
      pacparser.cleanup()
      return 1
    try:
      pacparser.parse_pac_string(script)
    except Exception:
      sys.stderr.write("pactester: Could not parse the pac script: %s\n"
                       % script)
      pacparser.cleanup()
      return 1
  else:
    try:
      pacparser.parse_pac_file(pacfile)
    except Exception:
      sys.stderr.write("pactester: Could not parse the pac file: %s\n"
                       % pacfile)
      pacparser.cleanup()
      return 1

  if client_ip:
    pacparser.setmyip(client_ip)

  if url:
    if not host:
      host = get_host_from_url(url)
      if not host:
        pacparser.cleanup()
        return 1
    try:
      proxy = pacparser.find_proxy(url, host)
    except Exception:
      sys.stderr.write("pactester: Problem in finding proxy for %s.\n" % url)
      pacparser.cleanup()
      return 1
    print(proxy)
    pacparser.cleanup()
    return 0

  if urlslist:
    try:
      fp = open(urlslist, "r")
    except IOError:
      sys.stderr.write("pactester: Could not open urlslist: %s" % urlslist)
      pacparser.cleanup()
      return 1
    for line in fp:
      u = line.lstrip(" \t")
      # Skip comment lines, echoing them as the C pactester does.
      if u.startswith("#"):
        sys.stdout.write(u)
        continue
      fields = u.split(None, 1)
      if not fields:
        continue
      u = fields[0]
      host = get_host_from_url(u)
      if not host:
        continue
      try:
        proxy = pacparser.find_proxy(u, host)
      except Exception:
        sys.stderr.write("pactester: Problem in finding proxy for %s.\n" % u)
        pacparser.cleanup()
        return 1
      if proxy:
        print("%s : %s" % (u, proxy))
    fp.close()
    pacparser.cleanup()
    return 0

  pacparser.cleanup()
  return 0

if __name__ == '__main__':
  sys.exit(main())
